cmd/config/io.go

```diff
@@ -9,20 +9,20 @@ import (
 )
 
 // ReadFile returns OneBuildConfiguration file content as string.
-func ReadFile() (string, error) {
-	if _, err := os.Stat(OneBuildConfigFileName); os.IsNotExist(err) {
-		return "", errors.New("no '" + OneBuildConfigFileName + "' file found in current directory")
+func ReadFile(oneBuildConfigFile string) (string, error) {
```

Comment:
So Idea is –


Bind the flag value with `viper`

In `root.go`

```go
rootCmd.PersistentFlags().StringP("file", "f", "", "some description")
	_ = viper.BindPFlags(rootCmd.PersistentFlags())
```

In  `IO package/parse.go`

```go
func getCurrentWorkingDirectoryFilePath() (string, error) {
	path, err := os.Getwd()
	if err != nil {
		return "", err
	}
	return path, nil
}

func getAbsoluteConfigurationFilePath() (string, error) {
	specifiedFilePath := viper.GetString("file")
	if specifiedFilePath == "" {
		return getCurrentWorkingDirectoryFilePath()
	}
	return specifiedFilePath, nil
}

// ReadFile returns OneBuildConfiguration file content as string.
func ReadFile() (string, error) {
	path, err := getAbsoluteConfigurationFilePath()
	if err != nil {
		return "", errors.New("error in reading '" + path + "' configuration file")
	}
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return "", errors.New("no '" + path + "' file found in current directory")
	}
	yamlFile, err := ioutil.ReadFile(path)
	if err != nil {
		return "", errors.New("error in reading '" + path + "' configuration file")
	}
	return string(yamlFile), nil
}
```


Everywhere else there will not able any use of which path is getting used.


Let me know if you need more clarification.

Reply:
@gopinath-langote 
Are you suggesting that I should keep within the existing i/o abstraction and things like `filepath` functions should be exported functions that other packages (for example `cmd/exec/exec.go`) can use? 

So, places like
https://github.com/gopinath-langote/1build/blob/8fb91017f9fd2b6d9a2f09a7c9d4fb1be10e26b9/cmd/exec/exec.go#L107

Should look like 

```go
// cmd/root.go
// Also notice the default value is set to global `config.OneBuildConfigFileName`
var FileFlag string
rootCmd.PersistentFlags().StringVarP(&FileFlag, "file", "f", config.OneBuildConfigFileName, "The File Path for 1build configuration file.")
_ = viper.BindPFlags(rootCmd.PersistentFlags())

// config/io.go
func GetAbsoluteDirFromConfigFile()  string {
        oneBuildConfig := viper.GetString("file")
	abs, err := filepath.Abs(oneBuildConfig)
	if err != nil {
		utils.CPrintln("\nError resolving file path for 1build configuration file.",
		utils.Style{Color: utils.RED, Bold: true})
		utils.ExitError()
	}
	baseDirFromAbs := filepath.Dir(abs)
	return baseDirFromAbs
}
// exec/exec.go
// **notice the removal of `(configFilePath string)` function param.**
func bashCommand(s *sh.Session, command string) *sh.Session {
	onebuildDir := config.GetAbsoluteDirFromConfigFile()
	s.SetDir(onebuildDir)
	return s.Command("bash", "-c", command)
}
```

Reply:
Yes. This is my suggestion- abstract out which file is getting used to io package. Also soon we will deprecate global  `config.OneBuildConfigFileName`

so avoid use if that variable instead use function to get path.

Reply:
@gopinath-langote I pushed a change that abstracted most of the config file, and now most of that logic resides inside `io.go / parse.go`.
https://github.com/gopinath-langote/1build/pull/194/commits/6f601f4961097d43d87e1bcc4fd8ece5c057bd63